### PR TITLE
Simple RDNA3 emulator

### PR DESCRIPTION
**Pull request by Qazalin**
https://github.com/Qazalin/remu interfacing with tinygrad. It has infrastructure to emulate RDNA3 execution, hardware state and _single-threaded_ kernel launch.

### CODE REVIEW

File: tinygrad/runtime/ops_hip.py [hunk 1]
```diff
@@ -1,13 +1,25 @@
...
+  remu = ctypes.CDLL("/usr/local/lib/libremu.so")
+  # TODO this whole section will be removed, the api should ideally match 1:1
+  hip.hipSetDevice = lambda x: x
+  hip.hipDeviceptr_t = lambda: ctypes.c_void_p(None)
+  hip.hipModule_t = lambda: ctypes.c_void_p(None)
+  hip.hipDeviceSynchronize = lambda: 0
+  hip.hipMalloc = remu.hipMalloc
+  hip.hipMemcpy = remu.hipMemcpy
+  hip.hipModuleLaunchKernel.argtypes = [ctypes.c_char_p, ctypes.c_int, ctypes.c_int, ctypes.c_int, ctypes.c_int, ctypes.c_int, ctypes.c_int, ctypes.c_int, ctypes.c_int, ctypes.c_void_p, ctypes.c_void_p, ctypes.c_uint, ctypes.c_void_p]  # noqa: E501
+  hip.hipModuleLaunchKernel = lambda prg, *args: remu.hipModuleLaunchKernel(prg, len(prg), *args)
```
Comment (g1y5x3) [inline tinygrad/runtime/ops_hip.py:22]: These technically belongs to gpuctypes, ideally just do import gpuctypes.hip as hip, hipcpu?

**Comment by unknownusername504** [comment]
Is this a functional pull request? Seems like a duplicate of this repo ... https://github.com/unknownusername504/RDNA3EMU

Should efforts on this be stopped? The two are similar but unmergeable.

File: test/test_dtype_alu.py [hunk 2]
```diff
@@ -143,9 +143,10 @@ def test_bool(self, a, b, op): universal_test(a, b, dtypes.bool, op)
   @given(ht.int32, ht.int32, ht.float32, strat.sampled_from(integer_binary_operations), strat.sampled_from(binary_operations))
   def test_int32_midcast_float(self, a, b, c, op1, op2): universal_test_midcast(a, b, c, op1, op2, dtypes.int32, dtypes.float32)
 
-  # Metal and CUDACPU behave differently than numpy in CI for overflows
-  @given(strat.floats(width=32, min_value=0, max_value=10.0) if CI and (Device.DEFAULT == "METAL" or getenv("CUDACPU")) else ht.float32,
-         strat.floats(width=32, min_value=0, max_value=10.0) if CI and (Device.DEFAULT == "METAL" or getenv("CUDACPU")) else ht.float32,
+  # Metal and CUDACPU and HIP behave differently than numpy in CI for overflows
```
Comment (Qazalin) [inline test/test_dtype_alu.py:146]: This test fails on real HIP with the same error. I think it's something HIP-specific with how they generate the assembly.
image

**Comment by geohot** [comment]
Why FORWARD_ONLY?

**Comment by Qazalin** [comment]
because of speed, [we also do it for CUDACPU.](https://github.com/tinygrad/tinygrad/blob/master/.github/workflows/test.yml#L405)
Screenshot 2024-01-29 at 3 17 08 PM

[FORWARD_ONLY=1](https://github.com/tinygrad/tinygrad/actions/runs/7696539117/job/20971590251):
Screenshot 2024-01-29 at 3 22 40 PM

Including [backward passes](https://github.com/tinygrad/tinygrad/actions/runs/7696648805/job/20971919935?pr=2974):
Screenshot 2024-01-29 at 3 21 41 PM

File: tinygrad/runtime/ops_hip.py [hunk 3]
```diff
@@ -172,3 +167,7 @@ def __call__(self, rawbufs:List[Buffer], var_vals, wait=False, jit=False):
     hip_set_device(self.device.device)
     check(hip.hipStreamWaitValue32(None, rawbufs[0]._buf, 1, 1, 0xFFFFFFFF))
     update_stats(colored("wait", "RED"), 0, 0, {}, None, 1, jit, device=self.dname)
+
+if getenv("HIPCPU"):
+  from extra.backends.ops_emulatedhip import EmulatedHIPDevice
```
Comment (geohot) [inline tinygrad/runtime/ops_hip.py:172]: This shouldn't import from extra, just put it in here

**Pull request merged**